Assets/Scripts/AchievementsService.cs

```diff
@@ -348,34 +353,46 @@ public Texture2D GetAchievementLockedIconTexture(string achievementId)
         /// The icon texture associated with the given achievement ID and
         /// determined by the given URI selector.
         /// </returns>
-        private Texture2D GetAchievementIconTexture(string achievementId, Func<DefinitionV2, string> uriSelector)
+        private Task<Texture2D> GetAchievementIconTexture(string achievementId, Func<DefinitionV2, string> uriSelector)
         {
-            Texture2D iconTexture = null;
+            TaskCompletionSource<Texture2D> tcs = new();
             foreach (var achievementDef in _achievements)
             {
                 if (achievementDef.AchievementId != achievementId)
                     continue;
 
-                if (_downloadCache.TryGetValue(uriSelector(achievementDef), out byte[] iconBytes) && null != iconBytes)
+                var uri = uriSelector(achievementDef);
+
+                
+                Texture2D textureFromBytes = new(2, 2);
+
+                if (!_downloadCache.ContainsKey(uri))
                 {
-                    Texture2D textureFromBytes = new(2, 2);
-                    if (textureFromBytes.LoadImage(iconBytes))
-                    {
-                        iconTexture = textureFromBytes;
-                    }
-                    else
+                    GetAndCacheData(uri, data =>
                     {
-                        // TODO: Deal with circumstances where image could not be loaded for some reason
-                    }
-                    break;
+                        if (data.result != UnityWebRequest.Result.Success)
+                        {
+                            Debug.Log($"Could not download achievement icon: {data.result}.");
+                            tcs.SetResult(textureFromBytes);
+                        }
+                    });
                 }
-                else
+
+                if (_downloadCache.TryGetValue(uri, out byte[] iconBytes) && null != iconBytes)
                 {
-                    // TODO: Deal with circumstance where icon bytes are not cached, or are null
+                    if (!textureFromBytes.LoadImage(iconBytes))
+                    {
+                        Debug.LogWarning($"Could not load achievement icon bytes.");
+                        tcs.SetResult(textureFromBytes);
```

Comment:
Should this set an exception, since there was some kind of error?

Reply:
Exceptions should typically be utilized in circumstances where execution cannot continue. In this case, execution can continue, because it is handled gracefully and will not cause a crash.

Reply:
Same approach as the earlier comment. Marking this conversation as resolved since it's a duplicate-ish thing.